mlserver/handlers/dataplane.py

```diff
@@ -85,53 +87,101 @@ async def infer(
         name: str,
         version: Optional[str] = None,
     ) -> InferenceResponse:
+        # need to cache the payload here since it
+        # will be modified in the context manager
+        if self._response_cache is not None:
+            cache_key = payload.json()
+
+        async with self._infer_contextmanager(name, version) as model:
+            payload = self._prepare_payload(payload, model)
+
+            if (
+                self._response_cache is not None
+                and model.settings.cache_enabled is not False
+            ):
+                cache_value = await self._response_cache.lookup(cache_key)
+                if cache_value != "":
+                    prediction = InferenceResponse.parse_raw(cache_value)
+                else:
+                    prediction = await model.predict(payload)
+                    # ignore cache insertion error if any
+                    await self._response_cache.insert(cache_key, prediction.json())
+            else:
+                prediction = await model.predict(payload)
+
+            # Ensure ID matches
+            prediction.id = payload.id
+            self._inference_middleware.response_middleware(prediction, model.settings)
+            return prediction
+
+    async def infer_stream(
+        self,
+        payloads: AsyncIterator[InferenceRequest],
+        name: str,
+        version: Optional[str] = None,
+    ) -> AsyncIterator[InferenceResponse]:
+        # TODO: Implement cache for stream
+        async with self._infer_contextmanager(name, version) as model:
+            async for payload in payloads:
+                break
+
+            payload = self._prepare_payload(payload, model)
+            payloads_decorated = self._payloads_decorator(payload, payloads, model)
+
+            async for prediction in model.predict_stream(payloads_decorated):
+                prediction.id = payload.id  # Ensure ID matches
+                self._inference_middleware.response_middleware(
+                    prediction, model.settings
+                )
+                yield prediction
+
+    def _prepare_payload(
+        self, payload: InferenceRequest, model: MLModel
+    ) -> InferenceRequest:
+        if payload.id is None:
+            payload.id = generate_uuid()
+
+        self._inference_middleware.request_middleware(payload, model.settings)
+        return payload
+
+    async def _payloads_decorator(
+        self,
+        payload: InferenceRequest,
+        payloads: AsyncIterator[InferenceRequest],
```

Comment:
should we rename these variables as `first_payload` and `subsequent_payloads`, also add a note in the code to explain the rational?